### PR TITLE
feat(behavior_path_planner): add PullOutPath member variables for ego predicted path generation

### PR DESCRIPTION
**Pull request by kyoichi-sugahara**
## Description

I added member variables for `PullOutPath`.
They will be used for ego predicted path generation in safety check

## Tests performed

Not applicable.

## Effects on system behavior

Not applicable.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/src/utils/start_planner/shift_pull_out.cpp [hunk 1]
```diff
@@ -155,17 +155,26 @@ std::vector<PullOutPath> ShiftPullOut::calcPullOutPaths(
     start_planner_utils::calcEndArcLength(s_start, forward_path_length, road_lanes, goal_pose);
   const double s_end = path_end_info.first;
   const bool path_terminal_is_goal = path_end_info.second;
-
   constexpr double RESAMPLE_INTERVAL = 1.0;
   PathWithLaneId road_lane_reference_path = utils::resamplePathWithSpline(
     route_handler.getCenterLinePath(road_lanes, s_start, s_end), RESAMPLE_INTERVAL);
+  const double shift_distance = s_end - s_start;
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/utils/start_planner/shift_pull_out.cpp:161]: this is arc length in road lane, better to use shifted path length

File: planning/behavior_path_planner/src/utils/start_planner/shift_pull_out.cpp [hunk 2]
```diff
@@ -155,17 +155,26 @@ std::vector<PullOutPath> ShiftPullOut::calcPullOutPaths(
...
+  const size_t shift_end_idx =
+    findNearestIndex(road_lane_reference_path.points, goal_pose.position);
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/utils/start_planner/shift_pull_out.cpp:163]: goal_pose is from the route and is not the same to shift end pose.
end_pose is `shift_end_pose_ptr` and the idx is `shift_line_idx.end_idx`

```
    // create shift line
    ShiftLine shift_line{};
    shift_line.start = start_pose;
    shift_line.end = *shift_end_pose_ptr;
    shift_line.end_shift_length = shift_length;
    path_shifter.addShiftLine(shift_line);
    path_shifter.setVelocity(0.0);  // initial velocity is 0
    path_shifter.setLongitudinalAcceleration(longitudinal_acc);
    path_shifter.setLateralAccelerationLimit(lateral_acc);

    const auto shift_line_idx = path_shifter.getShiftLines().front();
    if (!has_non_shifted_path && (shift_line_idx.end_idx - shift_line_idx.start_idx <= 1)) {
      candidate_paths.push_back(non_shifted_path);
      has_non_shifted_path = true;
      continue;
    }
```

**Comment by kosuke55** [comment]
thanks for the PR!!!

File: planning/behavior_path_planner/src/utils/start_planner/shift_pull_out.cpp (hunk 1 above)
Comment (kyoichi-sugahara) [reply planning/behavior_path_planner/src/utils/start_planner/shift_pull_out.cpp:161]: @kosuke55 
Thank you so much for your check :heart: 
I think my bad variable name causes confusion.
Here, I want to calculate ditance between shift_start pose and shift_end pose along the non_shifted path.
And in my understanding, non_shifted_path is same with road_reference_path, correct?
Also I mistalenly used goal pose not shift_end_pose....
In this case in my opinion,
1. I should change variable name to `non_shifted_path_arc_length`.
2. I should use `shift_end_pose_ptr->position` not `goal_pose.position` as you suggested another comment.
So here is my proposal to revised code. (will clean if my understanging is correct)
```
  // non_shifted_path for when shift length or pull out distance is too short
  const PullOutPath non_shifted_path = std::invoke([&]() {
    PullOutPath non_shifted_path{};
    // The assumption for this calculation of velocity and acceleration is to accelerate with
    // constant acceleration until the road velocity limit is reached at the shift end point.
    const auto & arc_position_non_shifted_path_end =
      getArcCoordinates(road_lanes, shift_end_pose_ptr->position);
    const double non_shifted_path_arc_length = arc_position_non_shifted_path_end.length - s_start;
    const size_t non_shifted_path_end_idx =
    findNearestIndex(road_lane_reference_path.points, shift_end_pose_ptr->position);
    const double terminal_velocity =
      road_lane_reference_path.points.at(non_shifted_path_end_idx).point.longitudinal_velocity_mps;
    const double time_to_end_pose = shift_distance / terminal_velocity;
    non_shifted_path.partial_paths.push_back(road_lane_reference_path);
    non_shifted_path.start_pose = start_pose;
    non_shifted_path.end_pose = start_pose;
    non_shifted_path.acceleration = 2 * shift_distance / std::pow(time_to_end_pose, 2);
    non_shifted_path.terminal_velocity = terminal_velocity;
    return non_shifted_path;
  });
 ```

Comment (kosuke55) [reply planning/behavior_path_planner/src/utils/start_planner/shift_pull_out.cpp:161]: @kyoichi-sugahara 

> And in my understanding, non_shifted_path is same with road_reference_path, correct?

yes, correct!

And sorry, I don't understand about the proposal code.
- Why shifted_end_pose_ptr is used for non_shifted_path?
- The end_pose of non_shifted_path matches the start_pose, but to what extent are you trying to safety check?
- When the ego is close to the center line, I think isExecutionRequest will be set to false in the future, but what role will non_shifted_path play? (Isn't it just a guard?)

File: planning/behavior_path_planner/src/utils/start_planner/geometric_pull_out.cpp [hunk 3]
```diff
@@ -80,15 +80,30 @@ boost::optional<PullOutPath> GeometricPullOut::plan(Pose start_pose, Pose goal_p
...
+  const double velocity = parallel_parking_parameters_.forward_parking_velocity;
 
   if (parameters_.divide_pull_out_path) {
     output.partial_paths = planner_.getPaths();
     // insert stop velocity to first arc path end
+    const double arc_length_on_path =
+      motion_utils::calcArcLength(output.partial_paths.front().points);
     output.partial_paths.front().points.back().point.longitudinal_velocity_mps = 0.0;
+    output.pairs_terminal_velocity_and_accel.push_back(
+      std::make_pair(velocity, velocity * velocity / (2 * arc_length_on_path)));
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/utils/start_planner/geometric_pull_out.cpp:92]: partial_paths size is two, but only one `pairs_terminal_velocity_and_accel` is ok?

File: planning/behavior_path_planner/src/utils/start_planner/geometric_pull_out.cpp [hunk 4]
```diff
@@ -80,15 +80,30 @@ boost::optional<PullOutPath> GeometricPullOut::plan(Pose start_pose, Pose goal_p
...
+
+    // Calculate the acceleration required to reach the forward parking velocity at the center of
+    // the path, assuming constant acceleration and deceleration.
+    const double arc_length_on_path = motion_utils::calcArcLength(combined_path.points);
+    const double acceleration = velocity * velocity / arc_length_on_path;
+    const double time_to_center = velocity / acceleration;
+    const double average_velocity = arc_length_on_path / (time_to_center * 2);
+    const double average_acceleration = average_velocity / (time_to_center * 2);
+    output.pairs_terminal_velocity_and_accel.push_back(
+      std::make_pair(average_velocity, average_acceleration));
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/utils/start_planner/geometric_pull_out.cpp:106]: combine path is without a stop point, but why assume deceleration?
and the second path has road_center_line so, `the center of the path` is too far like forward_path_length/2

File: planning/behavior_path_planner/src/utils/start_planner/geometric_pull_out.cpp [hunk 5]
```diff
@@ -80,15 +80,48 @@ boost::optional<PullOutPath> GeometricPullOut::plan(Pose start_pose, Pose goal_p
...
     output.partial_paths.front().points.back().point.longitudinal_velocity_mps = 0.0;
+    const double arc_length_on_path_front =
+      motion_utils::calcArcLength(output.partial_paths.front().points);
+    const double time_to_center = arc_length_on_path_front / (2 * velocity);
+    const double average_velocity = arc_length_on_path_front / (time_to_center * 2);
+    const double average_acceleration = average_velocity / (time_to_center * 2);
+    output.pairs_terminal_velocity_and_accel.push_back(
+      std::make_pair(average_velocity, average_acceleration));
+    const double arc_length_on_path_back =
+      motion_utils::calcArcLength(output.partial_paths.back().points);
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/utils/start_planner/geometric_pull_out.cpp:112]: `output.partial_paths.back().points` is not up to end pose but combined with road reference path

We need to calculate arc length up to PullOutPath.end_pose 

https://github.com/autowarefoundation/autoware.universe/blob/main/planning/behavior_path_planner/include/behavior_path_planner/utils/start_planner/pull_out_path.hpp#L31

```
struct PullOutPath
{
  std::vector<PathWithLaneId> partial_paths{};
  Pose start_pose{};
  Pose end_pose{};
};
```

Comment (kyoichi-sugahara) [reply planning/behavior_path_planner/src/utils/start_planner/geometric_pull_out.cpp]: @kosuke55 
fix with the [commit](https://github.com/autowarefoundation/autoware.universe/pull/4669/commits/e566b3d099faf7704f416dca895561e8e7f8f7cc)

Comment (kyoichi-sugahara) [reply planning/behavior_path_planner/src/utils/start_planner/geometric_pull_out.cpp]: finally...
https://github.com/autowarefoundation/autoware.universe/pull/4669/commits/81d756d7ef2732e748feef70bf59b82e6b0a7841

File: planning/behavior_path_planner/src/utils/start_planner/geometric_pull_out.cpp [hunk 6]
```diff
@@ -80,15 +80,48 @@ boost::optional<PullOutPath> GeometricPullOut::plan(Pose start_pose, Pose goal_p
...
+    |    /\               |    /\
+    |   /  \              |   /  \
+    |  /    \             |  /    \
+    | /      \            | /      \
+    |/________\_____ x    |/________\______ t
+    0  a_l/2  a_l         0    t_c 2*t_c
+    Notes:
+    a_l represents "arc_length_on_path_front"
+    t_c represents "time_to_center"
+    */
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/utils/start_planner/geometric_pull_out.cpp:101]: cool!!!

**Review by kosuke55 — APPROVED** [review]
thanks!!!!!!!!!!!!!!!!!!!! LGTM!

**Pull request merged**